src/Stratis.Bitcoin/Consensus/Rules/RuleContext.cs

```diff
@@ -67,25 +31,49 @@ public class RuleContext
         public bool SkipValidation { get; set; }
 
         /// <summary>The current tip of the chain that has been validated.</summary>
-        public ChainedHeader ConsensusTip { get; set; }
+        public ChainedHeader PreviousChainedHeader { get; set; }
+
+        public int PreviousHeight { get; set; }
 
-        public bool IsPoS
+        public RuleContext()
         {
-            get { return this.Stake != null; }
+            this.objectsBag = new Dictionary<Type, object>();
         }
 
-        public RuleContext()
+        public T Item<T>() where T : class
```

Comment:
How is this better than normal inheritance?

Reply:
Yeah I am not sure which is better, I decide on this approach as if we use inheritance each rule would have to cast the context (which is ok I guess) but then the engine will need to know what cobtext to creat.

Possibly inheritance is also an option.

Reply:
My vote is on inheritance. For me, this is not nice.

Reply:
My argument against inheritance is that this is more flexible, if a new rule needs to set something in the bag that will be used by a later rule this becomes possible with inheritance this is much harder.

Reply:
I don't see such a problem and I also don't consider this a cleaner solution. For example it has a problem if you want to add to thinks of the same type. I would certainly go for inheritance, but if you are unconvinced, please ask more people so that we have more opinions.